middleware/metrics.go

```diff
@@ -0,0 +1,218 @@
+package middleware
+
+// Ported from Goji's middleware, source:
+// https://github.com/zenazn/goji/tree/master/web/middleware
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+)
+
+// Metric is a composition of various statistics from a request, such as its
+// execution time and the response status.
+type Metric struct {
+	RequestID     string
+	Method        string
+	URL           string
+	RemoteAddr    string
+	Status        int
+	BytesWritten  int
+	ExecutionTime time.Duration
+}
+
+// MetricsHandler is a Metric receiver.
+type MetricsHandler interface {
+	Add(metric Metric)
+}
+
+// Metrics return a middleware which will use the given MetricsHandler in
+// order to publish new statistics.
+func Metrics(handler MetricsHandler) func(next http.Handler) http.Handler {
```

Comment:
Hello,

It's a terminology choice in order to separate what the data is (a metric) with how it's handled (with a logger).

For example, depending of your use case, you may use this middleware to expose these metrics on one of these _"backends"_:

* NSQ Producer
* RabbitMQ Queue
* Telegraf (InfluxDB) Plugin
* Prometheus Exporter

With these examples, is Logger still make sense ? I was trying to "_use_" a more generic terms _(Metrics)_ for this purpose. _(However, a Logger use these metrics)_

Nevertheless, If the contributors team isn't conviced by this opinionated design, I can update this Pull Request with the terminology that you prefer.